curp/src/client.rs

```diff
@@ -1 +1,139 @@
+use std::{fmt::Debug, marker::PhantomData, net::SocketAddr, sync::Arc, time::Duration};
 
+use futures::{stream::FuturesUnordered, StreamExt};
+use madsim::net::Endpoint;
+use tracing::{error, log::warn};
+
+use crate::{
+    cmd::Command,
+    message::{Propose, ProposeError},
+};
+
+/// Propose request default timeout
+static PROPOSE_TIMEOUT: Duration = Duration::from_secs(1);
+
+/// Protocol client
+pub struct Client<C: Command> {
+    /// Client endpoint to send requests
+    ep: Arc<Endpoint>,
+    /// Leader address
+    #[allow(dead_code)] // should use in propose synced
+    leader: SocketAddr,
+    /// All servers addresses including leader address
+    addrs: Vec<SocketAddr>,
+    /// To keep Command type
+    phatom: PhantomData<C>,
+}
+
+impl<C> Debug for Client<C>
+where
+    C: Command,
+{
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Client")
+            .field("addrs", &self.addrs)
+            .field("phatom", &self.phatom)
+            .finish()
+    }
+}
+
+impl<C> Client<C>
+where
+    C: Command + 'static,
+{
+    /// Create a new protocol client based on the addresses
+    #[inline]
+    pub async fn new(leader: &SocketAddr, addrs: &[SocketAddr]) -> Option<Self> {
+        let ep = Endpoint::bind("127.0.0.1:0")
```

Comment:
Will this endpoint use random port number?

Reply:
Yes, port 0 means pick a random port number, it's the client end.